### PR TITLE
[ricos-plugins] plugin text buttons

### PR DESCRIPTION
**Pull request by noamtwix**

### CODE REVIEW

File: packages/ricos-plugins/web/src/editorPlugins.ts [hunk 1]
```diff
@@ -61,6 +62,15 @@ export class EditorPlugins implements IEditorPlugins {
     return this.plugins.forEach(plugin => plugin.configure(config));
   }
 
+  getTextButtons() {
+    //maybe use filter class func
+    const textButtons = this.plugins.reduce(
+      (prev, curr) => [...prev, ...(curr.getTextButtons() || [])],
+      []
+    );
+    return textButtons.map(button => new PluginTextButton(button.getButton()));
```
Comment (danarish4) [inline packages/ricos-plugins/web/src/editorPlugins.ts:71]: not necessary -> new PluginTextButton(button.getButton()) === button
```suggestion
    return textButtons;
```

File: packages/ricos-plugins/web/src/pluginTextButton.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import type { IPluginTextButton, IToolbarItemConfigTiptap } from 'ricos-types';
```
Comment (danarish4) [inline packages/ricos-plugins/web/src/pluginTextButton.tsx:1]: Nothing is done in this class. It seems unnecessary to me

File: packages/toolbars-v3/web/src/toolbarItemConfig/tiptapToolbarItemConfig.ts [hunk 3]
```diff
@@ -265,26 +265,26 @@ export const tiptapStaticToolbarConfig: IToolbarItemConfigTiptap[] = [
...
-    },
-    commands: {
-      toggleSpoiler:
-        ({ editorCommands }) =>
-        () => {
-          editorCommands.chain().focus().toggleSpoiler().run();
-        },
-    },
-  },
+  // {
```
Comment (danarish4) [inline packages/toolbars-v3/web/src/toolbarItemConfig/tiptapToolbarItemConfig.ts:268]: remove comments

**Review by danarish4 — COMMENTED** [review]
I would remove `PluginTextButton` class for now and return plugin.textButtons

File: packages/ricos-plugins/web/src/editorPlugin.ts [hunk 4]
```diff
@@ -24,9 +27,16 @@ export class EditorPlugin implements IEditorPlugin {
...
     this.initAddButtons(plugin, modalService);
+    this.initTextButtons(plugin);
     this.initPluginToolbar(plugin, modalService);
   }
 
+  private initTextButtons(plugin: EditorPluginType) {
+    if (plugin.textButtons) {
+      this.textButtons = plugin.textButtons;
+    }
+  }
```
Comment (danarish4) [inline packages/ricos-plugins/web/src/editorPlugin.ts:38]:
```suggestion
```

no need to init anything

File: packages/ricos-plugins/web/src/editorPlugin.ts [hunk 5]
```diff
@@ -24,9 +27,16 @@ export class EditorPlugin implements IEditorPlugin {
   private constructor(plugin: EditorPluginType, modalService?: ModalService) {
     this.plugin = plugin;
     this.initAddButtons(plugin, modalService);
+    this.initTextButtons(plugin);
```
Comment (danarish4) [inline packages/ricos-plugins/web/src/editorPlugin.ts:30]:
```suggestion
```

File: packages/ricos-plugins/web/src/editorPlugin.ts [hunk 6]
```diff
@@ -72,6 +82,10 @@ export class EditorPlugin implements IEditorPlugin {
     return this.plugin.config;
   }
 
+  getTextButtons() {
+    return this.textButtons || [];
```
Comment (danarish4) [inline packages/ricos-plugins/web/src/editorPlugin.ts:86]:
```suggestion
    return this.plugin.textButtons || [];
```

File: packages/ricos-plugins/web/src/editorPlugin.ts [hunk 7]
```diff
@@ -15,6 +16,8 @@ export class EditorPlugin implements IEditorPlugin {
 
   addButtons?: IPluginAddButton[];
 
+  textButtons?: IToolbarItemConfigTiptap[];
```
Comment (danarish4) [inline packages/ricos-plugins/web/src/editorPlugin.ts:19]:
```suggestion
```

**Review by danarish4 — CHANGES_REQUESTED** [review]
follow comments :)

File: packages/ricos-plugins/web/src/editorPlugins.ts [hunk 8]
```diff
@@ -61,6 +61,15 @@ export class EditorPlugins implements IEditorPlugins {
...
   }
 
+  getTextButtons() {
+    //maybe use filter class func
+    const textButtons = this.plugins.reduce(
+      (prev, curr) => [...prev, ...(curr.getTextButtons() || [])],
+      []
+    );
+    return textButtons;
+  }
```
Comment (lxgreen) [inline packages/ricos-plugins/web/src/editorPlugins.ts:71]: What is the point to expose `IToolbarItemConfigTiptap[]`? What's its responsibility? 

Hint: who's responsible to order buttons accordingly to consumer config?

File: packages/ricos-plugins/web/src/editorPlugin.ts [hunk 9]
```diff
@@ -72,6 +72,10 @@ export class EditorPlugin implements IEditorPlugin {
     return this.plugin.config;
   }
 
+  getTextButtons() {
+    return this.plugin.textButtons || [];
+  }
+
```
Comment (lxgreen) [inline packages/ricos-plugins/web/src/editorPlugin.ts:78]: What is the point to expose `IToolbarItemConfigTiptap[]`? What's its responsibility?

**Review by lxgreen — CHANGES_REQUESTED** [review]

File: packages/ricos-types/web/src/editorPluginTypes.ts [hunk 10]
```diff
@@ -43,6 +43,13 @@ export interface IEditorPlugin {
    * @memberof IEditorPlugin
    */
   getConfig(): LegacyEditorPluginConfig;
+  /**
```
Comment (lxgreen) [inline packages/ricos-types/web/src/editorPluginTypes.ts:46]: @noamtwix @danarish4 
please make sure that `ricos-types` contains public types only. seems like you put here all the model types that should be defined in `ricos-plugins` only.

File: packages/ricos-plugins/web/src/editorPlugins.ts [hunk 11]
```diff
@@ -61,6 +62,15 @@ export class EditorPlugins implements IEditorPlugins {
     return this.plugins.forEach(plugin => plugin.configure(config));
   }
 
+  getTextButtons() {
+    //maybe use filter class func
```
Comment (lxgreen) [inline packages/ricos-plugins/web/src/editorPlugins.ts:66]: what's this comment for?

File: packages/ricos-plugins/web/src/editorPlugins.ts [hunk 12]
```diff
@@ -61,6 +62,15 @@ export class EditorPlugins implements IEditorPlugins {
...
+    const textButtons = this.plugins.reduce(
+      (prev, curr) => [...prev, ...(curr.getTextButtons() || [])],
+      []
+    );
```
Comment (lxgreen) [inline packages/ricos-plugins/web/src/editorPlugins.ts:70]: see `Array.prototype.flatMap`

File: packages/ricos-plugins/web/src/pluginTextButton.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+export class PluginTextButton implements IPluginTextButton {
+  button: IToolbarItemConfigTiptap;
+
+  constructor(button: IToolbarItemConfigTiptap) {
+    this.button = button;
+  }
+
+  static of(button: IToolbarItemConfigTiptap): IPluginTextButton {
+    return new PluginTextButton(button);
+  }
```
Comment (lxgreen) [inline packages/ricos-plugins/web/src/pluginTextButton.tsx:21]: what's the point to define both static `of` and public constructor? the `of` supposed to hide construction in some cases. 
Also, `of` (if used) supposed to be a public method of **aggregate root**. Your root here is `EditorPlugins` class. All the models are internal relative to it.

File: packages/ricos-plugins/web/src/pluginTextButton.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
+import type { IPluginTextButton, IPluginTextButtons, IToolbarItemConfigTiptap } from 'ricos-types';
```
Comment (lxgreen) [inline packages/ricos-plugins/web/src/pluginTextButton.tsx:1]: please stop using `I` prefix for interfaces

**Review by lxgreen — COMMENTED** [review]
1. file names
2. public types only in `ricos-types`
3. class methods
4. unused code

**Pull request merged**